lib/block_view.go

```diff
@@ -2919,24 +3053,72 @@ func _isEntryImmatureBlockReward(utxoEntry *UtxoEntry, blockHeight uint32, param
 	return false
 }
 
-func _verifySignature(txn *MsgBitCloutTxn) error {
-	// Compute a hash of the transaction
+func (bav *UtxoView) _verifySignature(txn *MsgBitCloutTxn, blockHeight uint32) error {
+	// Compute a hash of the transaction.
 	txBytes, err := txn.ToBytes(true /*preSignature*/)
 	if err != nil {
 		return errors.Wrapf(err, "_verifySignature: Problem serializing txn without signature: ")
 	}
 	txHash := Sha256DoubleHash(txBytes)
-	// Convert the txn public key into a *btcec.PublicKey
-	txnPk, err := btcec.ParsePubKey(txn.PublicKey, btcec.S256())
+	sig, err := btcec.ParseDERSignature(txn.Signature, btcec.S256())
+
+	// Look for the derived key in transaction ExtraData and validate it. For transactions
+	// signed using a derived key, the derived public key is passed to ExtraData.
+	var derivedPk *btcec.PublicKey
+	var derivedPkBytes []byte
+	isDerived := false
+	if txn.ExtraData != nil {
+		if derivedPkBytes, isDerived = txn.ExtraData[DerivedPublicKey]; isDerived {
+			derivedPk, err = btcec.ParsePubKey(derivedPkBytes, btcec.S256())
+			if err != nil {
+				return RuleErrorDerivedKeyInvalidExtraData
+			}
+		}
+	}
+
+	// Get the owner public key and attempt turning it into *btcec.PublicKey.
+	ownerPkBytes := txn.PublicKey
+	ownerPk, err :=  btcec.ParsePubKey(ownerPkBytes, btcec.S256())
 	if err != nil {
-		return errors.Wrapf(err, "_verifySignature: Problem parsing public key: ")
+		return errors.Wrapf(err, "_verifySignature: Problem parsing owner public key: ")
 	}
-	// Verify that the transaction is signed by the specified key.
-	if txn.Signature == nil || !txn.Signature.Verify(txHash[:], txnPk) {
-		return RuleErrorInvalidTransactionSignature
+
+	// If no derived key is present in ExtraData, we check if transaction was signed by the owner.
+	// If derived key is present in ExtraData, we check if transaction was signed by the derived key.
+	if isDerived == false {
```

Comment:
Nit: isDerived is redundant and error-prone. I would instead do the following:

1) Define isDerived as a local variable in the `if txn.ExtraData != nil` block

2) Instead of checking `isDerived == false` do `derivedPk == nil`